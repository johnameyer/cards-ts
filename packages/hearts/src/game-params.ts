import { Serializable } from '@cards-ts/core';

export interface GameParams {
    readonly [key: string]: Serializable;
    
    readonly maxScore: number;

    readonly numToPass: number;

    readonly quickEnd: boolean;
}