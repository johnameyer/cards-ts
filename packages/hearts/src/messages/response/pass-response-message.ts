import { Card, Message } from '@cards-ts/core';

/**
 * Passes a certain number of cards and optionally sets the data for this player
 */
export class PassResponseMessage extends Message {
    readonly type = 'pass-response';

    constructor(public readonly cards: Card[], public readonly data?: any) {
        super([ 'Passed', cards ]);
    }
}
