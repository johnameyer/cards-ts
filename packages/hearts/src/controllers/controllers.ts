import { GameStates } from '../game-states';
import { GameParams } from '../game-params';
import { GameHandlerParams } from '../game-handler-params';
import { ResponseMessage } from '../messages/response-message';
import { PassingControllerProvider } from './passing-controller';
import { DefaultControllerKeys, DefaultControllers, SystemHandlerParams, DeckControllerProvider, HandsControllerProvider, PointsControllerProvider, TricksControllerProvider, TurnControllerProvider, ControllersProviders, UnwrapProviders, ValidatedProviders } from '@cards-ts/core';

type TypedDefaultControllers = DefaultControllers<GameParams, typeof GameStates, ResponseMessage, GameHandlerParams & SystemHandlerParams>;

export const buildProviders = () => {
    const providers = {
        deck: new DeckControllerProvider(1, true, false),
        hand: new HandsControllerProvider(),
        trick: new TricksControllerProvider(),
        trickPoints: new PointsControllerProvider(),
        passing: new PassingControllerProvider(),
        score: new PointsControllerProvider(),
        turn: new TurnControllerProvider(),
    };
    return providers as Omit<ValidatedProviders<typeof providers & ControllersProviders<TypedDefaultControllers>>, DefaultControllerKeys>;
};

export type GameControllers = UnwrapProviders<ReturnType<typeof buildProviders>>;

export type Controllers = GameControllers & TypedDefaultControllers;
