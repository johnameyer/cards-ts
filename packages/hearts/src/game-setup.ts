import { GenericGameSetup, Intermediary } from '@cards-ts/core';
import { GameParams } from './game-params';

export class GameSetup implements GenericGameSetup<GameParams> {
    getDefaultParams(): GameParams {
        return {
            maxScore: 100,
            numToPass: 3,
            quickEnd: true,
        };
    }
    
    async setupForIntermediary(host: Intermediary): Promise<GameParams> {
        const [ _, resultsPromise ] = host.form(
            { type: 'input', message: [ 'Score to play to? (default 100)' ] },
            { type: 'input', message: [ 'Number of cards to pass? (default 3)' ] },
            { type: 'confirm', message: [ 'End a round quickly if no points are remaining?' ] },
        );

        const results = await resultsPromise;

        const maxScore = Number(results[0]) || 100;
        const numToPass = Number(results[1]) || 3;
        const quickEnd = results[2];

        return {
            maxScore,
            numToPass,
            quickEnd,
        };
    }

    verifyParams(params: GameParams) {
        const errors: { maxScore?: string; numToPass?: string; quickEnd?: string; } = {};
        try {
            if(params.quickEnd !== true && params.quickEnd !== false) {
                throw new Error();
            }
        } catch (e) {
            errors.quickEnd = 'Quick-end must be a boolean.';
        }
        try {
            if(!Number(params.maxScore) || Number(params.maxScore) <= 0) {
                throw new Error();
            }
        } catch (e) {
            errors.maxScore = 'Max score must be a number greater than 0';
        }
        try {
            if(!Number(params.numToPass) || Number(params.numToPass) < 0 || Number(params.numToPass) > 4) {
                throw new Error();
            }
        } catch (e) {
            errors.numToPass = 'Number to pass must be between 0 and 4.';
        }
        return errors;
    }

    getYargs(): {[key: string]: import('yargs').Options} {
        return {
            quickEnd: { description: 'End a round quickly if no points are remaining', type: 'boolean', default: true },
            maxScore: { description: 'Score to play to', type: 'number', default: 100 },
            numberToPass: { description: 'Number of cards to pass', type: 'number', default: 3 },
        };
    }

    setupForYargs(params: any): GameParams {
        return {
            quickEnd: !!params.quickEnd,
            maxScore: Number(params.maxScore),
            numToPass: Number(params.numberToPass),
        };
    }
}
