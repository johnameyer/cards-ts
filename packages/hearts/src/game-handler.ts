import { Handler, HandlerAction } from "@cards-ts/core";
import { HandlerData } from "./handler-data";
import { DataResponseMessage, PassResponseMessage, TurnResponseMessage } from "./messages/response";
import { ResponseMessage } from "./messages/response-message";

export type GameHandlerParams = {
    pass: [],
    turn: []
}

export abstract class GameHandler implements Handler<GameHandlerParams, HandlerData, ResponseMessage> {
    abstract handlePass: HandlerAction<HandlerData, DataResponseMessage | PassResponseMessage>;

    abstract handleTurn: HandlerAction<HandlerData, DataResponseMessage | TurnResponseMessage>;
}