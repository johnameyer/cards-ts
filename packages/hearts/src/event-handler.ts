import { Suit, Card, Rank, distinct, isDefined, EventHandlerInterface, GenericGameState, PlayCardResponseMessage } from "@cards-ts/core";
import { Controllers } from "./controllers/controllers";
import { PassResponseMessage } from "./messages/response";
import { ResponseMessage } from "./messages/response-message";

const QS = new Card(Suit.SPADES, Rank.QUEEN);

export class EventHandler implements EventHandlerInterface<Controllers, ResponseMessage> {
    validateEvent(controllers: Controllers, source: number, event: ResponseMessage): ResponseMessage | undefined {
        switch(event.type) {
            case 'pass-response': {
                const { cards, data } = event;
                // console.log(cards.toString());
                // console.log(controllers.hand.get(source).sort(compare).toString());
                try {
                    if(!cards.every(isDefined)) {
                        // TODO better condition
                        throw new Error('Not all are cards');
                    }

                    if(cards.length !== controllers.params.get().numToPass) {
                        throw new Error('Wrong number of cards passed');
                    }

                    if(cards.filter(distinct).length !== controllers.params.get().numToPass) {
                        throw new Error('Cannot pass same card multiple times');
                    }

                    if(cards.some(card => controllers.hand.get(source).find(card.equals.bind(card)) === undefined)) {
                        throw new Error('Can only pass cards that are in hand');
                    }

                    return new PassResponseMessage(cards, data);
                } catch (e) {
                    console.error('Invalid pass', e);
                }

                return new PassResponseMessage(controllers.hand.get(source).slice(0, controllers.params.get().numToPass), data);
            }
            case 'turn-response': {
                if(source !== controllers.turn.get()) {
                    return undefined;
                }
                const { card, data } = event;
                // console.log(card.toString());
                // console.log(controllers.hand.get(source).toString());
                try {
                    if(!card) {
                        throw new Error('No card provided');
                    }

                    if(!controllers.hand.hasCard(card, source)) {
                        throw new Error('Cannot play card that is not in hand');
                    }
                    
                    if(controllers.trick.tricks === 0 && (card.suit === Suit.HEARTS || card.equals(QS))) {
                        throw new Error('Cannot give points on the first round');
                    }
        
                    if(controllers.trick.currentTrick[0] === undefined && card.suit === Suit.HEARTS && controllers.trickPoints.get().every(point => point === 0)) {
                        throw new Error('Blood has not been shed yet');
                    }
        
                    if(controllers.trick.currentTrick[0]?.suit && card.suit !== controllers.trick.currentTrick[0].suit && controllers.hand.get(source).some(card => card.suit === controllers.trick.currentTrick[0]?.suit)) {
                        throw new Error('Must follow suit if possible');
                    }

                    return new PlayCardResponseMessage(card, data);
                } catch (e) {
                    console.error('Invalid turn', e);
                }

                const fallbackCard = controllers.hand.get(source).filter(card => card.suit === controllers.trick.currentTrick[0]?.suit)[0] || controllers.hand.get(source).filter(card => card.suit !== Suit.HEARTS)[0] || controllers.hand.get(source)[0];
                return new PlayCardResponseMessage(fallbackCard, data);
            }
            case 'data-response': {
                return event;
            }
        }
    }

    
    merge(controllers: Controllers, sourceHandler: number, incomingEvent: ResponseMessage): void {
        switch(incomingEvent.type) {
            case 'pass-response': {
                controllers.passing.setPassedFor(sourceHandler, incomingEvent.cards);
                controllers.data.setDataFor(sourceHandler, incomingEvent.data);
                controllers.waiting.removePosition(sourceHandler);
                return;
            }
            case 'turn-response': {
                controllers.trick.setPlayedCard(incomingEvent.card);
                controllers.data.setDataFor(sourceHandler, incomingEvent.data);
                controllers.waiting.removePosition(sourceHandler);
                return;
            }
            case 'data-response': {
                controllers.data.setDataFor(sourceHandler, incomingEvent.data);
                return;
            }
        }
    }
}