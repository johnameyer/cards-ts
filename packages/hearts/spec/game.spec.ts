import { expect } from 'chai';
import { IntermediaryHandler } from '../src/index.js';
import { eventHandler } from '../src/event-handler.js';
import { GameSetup } from '../src/game-setup.js';
import { gameStateTransitions } from '../src/game-state-transitions.js';
import { GameHandler } from '../src/game-handler.js';
import { GameParams } from '../src/game-params.js';
import { buildProviders } from '../src/controllers/controllers.js';
import { HeuristicHandler } from '../src/handlers/heuristic-handler.js';
import { PassResponseMessage } from '../src/messages/response/pass-response-message.js';
import { PassingMessage } from '../src/messages/status/passing-message.js';
import { valueOfCard } from '../src/util/value-of-card.js';
import { followsTrick } from '../src/util/follows-trick.js';
import { PlayedMessage } from '../src/messages/status/played-message.js';
import { ArrayMessageHandler, buildGameFactory, Card, DeckControllerProvider, EndRoundMessage, HandlerChain, LeadsMessage, Message, PlayCardResponseMessage } from '@cards-ts/core';
import { StatusMessage } from '../../can-i-have-that/src/messages/status-message.js';

describe('game', () => {
    // TODO can we build this more simply i.e. deterministic deck controller?
    const mockProviders = () => ({
        ...buildProviders(),
        deck: new DeckControllerProvider(1, false, false), // note shouldShuffle false
    });

    const factory = buildGameFactory(
        gameStateTransitions,
        eventHandler,
        // TODO can we avoid specifying the setup and intermediary since they are not critical to the driver construction?
        new GameSetup(),
        intermediary => new IntermediaryHandler(intermediary),
        () => new HeuristicHandler(),
        mockProviders,
    );

    const params: GameParams = {
        ...new GameSetup().getDefaultParams(),
        maxScore: 26,
    };

    it('works as expected', async () => {
        const messageHandlers = Array.from({length: 4}, () => new ArrayMessageHandler<StatusMessage>());

        const gameHandler: () => GameHandler = () => ({
            handlePass: (state, responses) => {
                responses.push(new PassResponseMessage(state.hand.slice(0, 3)));
            },
            handleTurn: (state, responses) => {
                // TODO create DeterministicHandler (simpler logic than heuristic handler)
                const followingTrick = state.hand.filter(card => followsTrick(state.trick.currentTrick, card, state.hand));
                if(state.trickPoints.every(points => points === 0)) {
                    const pointless = (followingTrick.length ? followingTrick : state.hand).filter(card => valueOfCard(card) === 0);
                    responses.push(new PlayCardResponseMessage((pointless.length ? pointless : followingTrick)[0]));
                }
                responses.push(new PlayCardResponseMessage((followingTrick.length ? followingTrick : state.hand)[0]));
            },
        });

        // @ts-expect-error
        const players = Array.from(messageHandlers, messageHandler => new HandlerChain([ messageHandler, gameHandler() ]));

        const names = Array.from(messageHandlers, (_, i) => String.fromCharCode(65 + i));

        const driver = factory.getGameDriver(players, params, names);

        driver.resume();

        for(const messageHandler of messageHandlers) {
            expect(messageHandler.arr[0].type).to.equal('dealt-out-message');
            expect(messageHandler.arr[1]).to.deep.equal(new PassingMessage(3, 1, 4));

            messageHandler.clear();
        }

        expect(driver.getState().completed).to.be.false;

        driver.handleSyncResponses();
        driver.resume();

        for(const i in messageHandlers) {
            const messageHandler = messageHandlers[i];

            expect(messageHandler.arr[0].type).to.equal('passed-message');
            expect(messageHandler.arr[1]).to.deep.equal(new LeadsMessage('B'));

            if(i !== '1') { // B
                expect(messageHandler.arr[2]).to.deep.equal(new PlayedMessage('B', Card.fromString('2C')));
            }

            messageHandler.clear();
        }

        expect(driver.getState().completed).to.be.false;

        driver.handleSyncResponses();
        driver.resume();

        for(const i in messageHandlers) {
            const messageHandler = messageHandlers[i];
            if(i !== '2') { // C
                expect(messageHandler.arr[0]).to.deep.equal(new PlayedMessage('C', Card.fromString('3C', 0)));
            }

            messageHandler.clear();
        }

        expect(driver.getState().completed).to.be.false;

        driver.handleSyncResponses();
        driver.resume();

        for(const i in messageHandlers) {
            const messageHandler = messageHandlers[i];
            if(i !== '3') { // D
                expect(messageHandler.arr[0]).to.deep.equal(new PlayedMessage('D', Card.fromString('4C', 0)));
            }
            messageHandler.clear();
        }

        expect(driver.getState().completed).to.be.false;

        driver.handleSyncResponses();
        driver.resume();

        for(const i in messageHandlers) {
            const messageHandler = messageHandlers[i];
            if(i !== '0') { // A
                expect(messageHandler.arr[0]).to.deep.equal(new PlayedMessage('A', Card.fromString('5C', 0)));
            }
            messageHandler.clear();
        }

        expect(driver.getState().completed).to.be.false;

        while(!driver.getState().completed) {
            driver.handleSyncResponses();
            driver.resume();
        }

        for(const i in messageHandlers) {
            const messageHandler = messageHandlers[i];
            expect(messageHandler.arr.at(-1)?.type).to.equal('end-round-message');

            expect((messageHandler.arr.at(-1) as EndRoundMessage).scores.reduce((a, b) => a + b, 0) % 26).to.equal(0);
        }
    });
});
