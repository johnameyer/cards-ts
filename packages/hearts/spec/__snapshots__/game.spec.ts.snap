// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`game has a matching snapshot 1`] = `
Array [
  "Received 3 of Spades, 7 of Spades",
  "Passing 3 cards 1 to the right",
  "Received 3 of Diamonds, 7 of Diamonds, J of Diamonds from B",
  "B leads next trick",
  "B played 2 of Clubs",
  "C played J of Clubs",
  "D played K of Clubs",
  "D leads next trick",
  "D played 2 of Spades",
  "B played 8 of Spades",
  "C played K of Spades",
  "C leads next trick",
  "C played 5 of Diamonds",
  "D played A of Spades",
  "B played Q of Diamonds",
  "B leads next trick",
  "B played 6 of Clubs",
  "C played 3 of Clubs",
  "D played 5 of Clubs",
  "B leads next trick",
  "B played 10 of Clubs",
  "C played 7 of Clubs",
  "D played 9 of Clubs",
  "B leads next trick",
  "B played 4 of Diamonds",
  "C played K of Diamonds",
  "D played Q of Clubs",
  "C leads next trick",
  "C played 2 of Hearts",
  "D played 3 of Hearts",
  "B played K of Hearts",
  "B leads next trick",
  "B played A of Clubs",
  "C played A of Hearts",
  "D played 8 of Clubs",
  "B leads next trick",
  "B played 8 of Diamonds",
  "C played 9 of Diamonds",
  "D played J of Hearts",
  "C leads next trick",
  "C played 6 of Hearts",
  "D played 7 of Hearts",
  "B played 5 of Hearts",
  "D leads next trick",
  "D played 4 of Clubs",
  "B played Q of Spades",
  "C played 10 of Hearts",
  "D leads next trick",
  "D played 6 of Spades",
  "B played 4 of Spades",
  "C played 9 of Spades",
  "C leads next trick",
  "C played 5 of Spades",
  "D played 10 of Spades",
  "B played 9 of Hearts",
  "Received 0 points",
  "A: 0 B: 7 C: 1 D: 18",
  "Received 3 of Spades, 7 of Spades",
  "Passing 3 cards 1 to the left",
  "Received 5 of Diamonds, 9 of Diamonds, K of Diamonds from D",
  "B leads next trick",
  "B played 2 of Clubs",
  "C played J of Clubs",
  "D played Q of Clubs",
  "D leads next trick",
  "D played 2 of Spades",
  "B played 8 of Spades",
  "C played K of Spades",
  "C leads next trick",
  "C played 3 of Diamonds",
  "D played Q of Diamonds",
  "B played Q of Spades",
  "D leads next trick",
  "D played 3 of Hearts",
  "B played 5 of Hearts",
  "C played 2 of Hearts",
  "B leads next trick",
  "B played 5 of Clubs",
  "C played 3 of Clubs",
  "D played 4 of Clubs",
  "B leads next trick",
  "B played 6 of Clubs",
  "C played 7 of Clubs",
  "D played 8 of Clubs",
  "D leads next trick",
  "D played 4 of Diamonds",
  "B played A of Clubs",
  "C played J of Diamonds",
  "C leads next trick",
  "C played 7 of Diamonds",
  "D played 8 of Diamonds",
  "B played K of Hearts",
  "D leads next trick",
  "D played 7 of Hearts",
  "B played 9 of Hearts",
  "C played 6 of Hearts",
  "B leads next trick",
  "B played 9 of Clubs",
  "C played A of Hearts",
  "D played A of Spades",
  "B leads next trick",
  "B played 10 of Clubs",
  "C played 10 of Hearts",
  "D played J of Hearts",
  "Received 0 points",
  "A: 0 B: 18 C: 1 D: 33",
]
`;

exports[`game has a matching snapshot 2`] = `
Array [
  "Received 4 of Spades, K of Clubs",
  "Passing 3 cards 1 to the right",
  "Received 4 of Diamonds, 8 of Diamonds, Q of Diamonds from C",
  "B leads next trick",
  "C played J of Clubs",
  "D played K of Clubs",
  "A played A of Diamonds",
  "D leads next trick",
  "D played 2 of Spades",
  "A played J of Spades",
  "C played K of Spades",
  "C leads next trick",
  "C played 5 of Diamonds",
  "D played A of Spades",
  "A played 3 of Diamonds",
  "B leads next trick",
  "C played 3 of Clubs",
  "D played 5 of Clubs",
  "A played Q of Hearts",
  "B leads next trick",
  "C played 7 of Clubs",
  "D played 9 of Clubs",
  "A played J of Diamonds",
  "B leads next trick",
  "C played K of Diamonds",
  "D played Q of Clubs",
  "A played 10 of Diamonds",
  "C leads next trick",
  "C played 2 of Hearts",
  "D played 3 of Hearts",
  "A played 4 of Hearts",
  "B leads next trick",
  "C played A of Hearts",
  "D played 8 of Clubs",
  "A played 8 of Hearts",
  "B leads next trick",
  "C played 9 of Diamonds",
  "D played J of Hearts",
  "A played 7 of Diamonds",
  "C leads next trick",
  "C played 6 of Hearts",
  "D played 7 of Hearts",
  "A played 7 of Spades",
  "D leads next trick",
  "D played 4 of Clubs",
  "A played 6 of Diamonds",
  "C played 10 of Hearts",
  "D leads next trick",
  "D played 6 of Spades",
  "A played 3 of Spades",
  "C played 9 of Spades",
  "C leads next trick",
  "C played 5 of Spades",
  "D played 10 of Spades",
  "A played 2 of Diamonds",
  "Received 7 points",
  "A: 0 B: 7 C: 1 D: 18",
  "Received 4 of Spades, K of Clubs",
  "Passing 3 cards 1 to the left",
  "Received 5 of Clubs, 9 of Clubs, K of Clubs from A",
  "B leads next trick",
  "C played J of Clubs",
  "D played Q of Clubs",
  "A played A of Diamonds",
  "D leads next trick",
  "D played 2 of Spades",
  "A played J of Spades",
  "C played K of Spades",
  "C leads next trick",
  "C played 3 of Diamonds",
  "D played Q of Diamonds",
  "A played 10 of Diamonds",
  "D leads next trick",
  "D played 3 of Hearts",
  "A played 4 of Hearts",
  "C played 2 of Hearts",
  "B leads next trick",
  "C played 3 of Clubs",
  "D played 4 of Clubs",
  "A played Q of Hearts",
  "B leads next trick",
  "C played 7 of Clubs",
  "D played 8 of Clubs",
  "A played 8 of Hearts",
  "D leads next trick",
  "D played 4 of Diamonds",
  "A played 2 of Diamonds",
  "C played J of Diamonds",
  "C leads next trick",
  "C played 7 of Diamonds",
  "D played 8 of Diamonds",
  "A played 6 of Diamonds",
  "D leads next trick",
  "D played 7 of Hearts",
  "A played K of Diamonds",
  "C played 6 of Hearts",
  "B leads next trick",
  "C played A of Hearts",
  "D played A of Spades",
  "A played 9 of Diamonds",
  "B leads next trick",
  "C played 10 of Hearts",
  "D played J of Hearts",
  "A played 5 of Diamonds",
  "Received 11 points",
  "A: 0 B: 18 C: 1 D: 33",
]
`;

exports[`game has a matching snapshot 3`] = `
Array [
  "Received 5 of Spades, 9 of Spades",
  "Passing 3 cards 1 to the right",
  "Received 5 of Diamonds, 9 of Diamonds, K of Diamonds from D",
  "B leads next trick",
  "B played 2 of Clubs",
  "D played K of Clubs",
  "A played A of Diamonds",
  "D leads next trick",
  "D played 2 of Spades",
  "A played J of Spades",
  "B played 8 of Spades",
  "C leads next trick",
  "D played A of Spades",
  "A played 3 of Diamonds",
  "B played Q of Diamonds",
  "B leads next trick",
  "B played 6 of Clubs",
  "D played 5 of Clubs",
  "A played Q of Hearts",
  "B leads next trick",
  "B played 10 of Clubs",
  "D played 9 of Clubs",
  "A played J of Diamonds",
  "B leads next trick",
  "B played 4 of Diamonds",
  "D played Q of Clubs",
  "A played 10 of Diamonds",
  "C leads next trick",
  "D played 3 of Hearts",
  "A played 4 of Hearts",
  "B played K of Hearts",
  "B leads next trick",
  "B played A of Clubs",
  "D played 8 of Clubs",
  "A played 8 of Hearts",
  "B leads next trick",
  "B played 8 of Diamonds",
  "D played J of Hearts",
  "A played 7 of Diamonds",
  "C leads next trick",
  "D played 7 of Hearts",
  "A played 7 of Spades",
  "B played 5 of Hearts",
  "D leads next trick",
  "D played 4 of Clubs",
  "A played 6 of Diamonds",
  "B played Q of Spades",
  "D leads next trick",
  "D played 6 of Spades",
  "A played 3 of Spades",
  "B played 4 of Spades",
  "C leads next trick",
  "D played 10 of Spades",
  "A played 2 of Diamonds",
  "B played 9 of Hearts",
  "Received 1 points",
  "A: 0 B: 7 C: 1 D: 18",
  "Received 5 of Spades, 9 of Spades",
  "Passing 3 cards 1 to the left",
  "Received 3 of Diamonds, 7 of Diamonds, J of Diamonds from B",
  "B leads next trick",
  "B played 2 of Clubs",
  "D played Q of Clubs",
  "A played A of Diamonds",
  "D leads next trick",
  "D played 2 of Spades",
  "A played J of Spades",
  "B played 8 of Spades",
  "C leads next trick",
  "D played Q of Diamonds",
  "A played 10 of Diamonds",
  "B played Q of Spades",
  "D leads next trick",
  "D played 3 of Hearts",
  "A played 4 of Hearts",
  "B played 5 of Hearts",
  "B leads next trick",
  "B played 5 of Clubs",
  "D played 4 of Clubs",
  "A played Q of Hearts",
  "B leads next trick",
  "B played 6 of Clubs",
  "D played 8 of Clubs",
  "A played 8 of Hearts",
  "D leads next trick",
  "D played 4 of Diamonds",
  "A played 2 of Diamonds",
  "B played A of Clubs",
  "C leads next trick",
  "D played 8 of Diamonds",
  "A played 6 of Diamonds",
  "B played K of Hearts",
  "D leads next trick",
  "D played 7 of Hearts",
  "A played K of Diamonds",
  "B played 9 of Hearts",
  "B leads next trick",
  "B played 9 of Clubs",
  "D played A of Spades",
  "A played 9 of Diamonds",
  "B leads next trick",
  "B played 10 of Clubs",
  "D played J of Hearts",
  "A played 5 of Diamonds",
  "Received 0 points",
  "A: 0 B: 18 C: 1 D: 33",
]
`;

exports[`game has a matching snapshot 4`] = `
Array [
  "Received 6 of Spades, 10 of Spades",
  "Passing 3 cards 1 to the right",
  "Received 5 of Clubs, 9 of Clubs, K of Clubs from A",
  "B leads next trick",
  "B played 2 of Clubs",
  "C played J of Clubs",
  "A played A of Diamonds",
  "D leads next trick",
  "A played J of Spades",
  "B played 8 of Spades",
  "C played K of Spades",
  "C leads next trick",
  "C played 5 of Diamonds",
  "A played 3 of Diamonds",
  "B played Q of Diamonds",
  "B leads next trick",
  "B played 6 of Clubs",
  "C played 3 of Clubs",
  "A played Q of Hearts",
  "B leads next trick",
  "B played 10 of Clubs",
  "C played 7 of Clubs",
  "A played J of Diamonds",
  "B leads next trick",
  "B played 4 of Diamonds",
  "C played K of Diamonds",
  "A played 10 of Diamonds",
  "C leads next trick",
  "C played 2 of Hearts",
  "A played 4 of Hearts",
  "B played K of Hearts",
  "B leads next trick",
  "B played A of Clubs",
  "C played A of Hearts",
  "A played 8 of Hearts",
  "B leads next trick",
  "B played 8 of Diamonds",
  "C played 9 of Diamonds",
  "A played 7 of Diamonds",
  "C leads next trick",
  "C played 6 of Hearts",
  "A played 7 of Spades",
  "B played 5 of Hearts",
  "D leads next trick",
  "A played 6 of Diamonds",
  "B played Q of Spades",
  "C played 10 of Hearts",
  "D leads next trick",
  "A played 3 of Spades",
  "B played 4 of Spades",
  "C played 9 of Spades",
  "C leads next trick",
  "C played 5 of Spades",
  "A played 2 of Diamonds",
  "B played 9 of Hearts",
  "Received 18 points",
  "A: 0 B: 7 C: 1 D: 18",
  "Received 6 of Spades, 10 of Spades",
  "Passing 3 cards 1 to the left",
  "Received 4 of Diamonds, 8 of Diamonds, Q of Diamonds from C",
  "B leads next trick",
  "B played 2 of Clubs",
  "C played J of Clubs",
  "A played A of Diamonds",
  "D leads next trick",
  "A played J of Spades",
  "B played 8 of Spades",
  "C played K of Spades",
  "C leads next trick",
  "C played 3 of Diamonds",
  "A played 10 of Diamonds",
  "B played Q of Spades",
  "D leads next trick",
  "A played 4 of Hearts",
  "B played 5 of Hearts",
  "C played 2 of Hearts",
  "B leads next trick",
  "B played 5 of Clubs",
  "C played 3 of Clubs",
  "A played Q of Hearts",
  "B leads next trick",
  "B played 6 of Clubs",
  "C played 7 of Clubs",
  "A played 8 of Hearts",
  "D leads next trick",
  "A played 2 of Diamonds",
  "B played A of Clubs",
  "C played J of Diamonds",
  "C leads next trick",
  "C played 7 of Diamonds",
  "A played 6 of Diamonds",
  "B played K of Hearts",
  "D leads next trick",
  "A played K of Diamonds",
  "B played 9 of Hearts",
  "C played 6 of Hearts",
  "B leads next trick",
  "B played 9 of Clubs",
  "C played A of Hearts",
  "A played 9 of Diamonds",
  "B leads next trick",
  "B played 10 of Clubs",
  "C played 10 of Hearts",
  "A played 5 of Diamonds",
  "Received 15 points",
  "A: 0 B: 18 C: 1 D: 33",
]
`;
